### ISSUE TITLE
Generic `Provider` type for files manipulation are lacking convenience methods

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle#3257]
`Directory` and `RegularFile` as properties support additional method to construct more of the same.

### Expected Behavior
`DirectoryProvider` and `RegularFileProvider` would offer the [`dir`](https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/file/DirectoryProperty.java#L49-L63) and [`file`](https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/file/DirectoryProperty.java#L65-L79) factory method to create providers based on the value of the root provider. The following would now be possible:
`someProvider.set(aDirectoryProvider.dir("child-folder"));`

### Current Behavior
At the moment, we have to go through the verbose `map` method as follow:
```
someProvider.set(aDirectoryProvider.map(new Transformer<Directory, Directory>() {
                    @Override
                    public Directory transform(Directory directory) {
                        return directory.dir("child-folder");
                    }
                }));
```

**Comment by ldaley** [comment]
FWIW, I think this is a bad road that we will regret. 

Having to create glue type specialisations for each type of actual thing is going to create an inconsistent experience because inevitably it will be patchy, increase the width of the API and ultimately make Gradle harder to learn. I also think it's likely that these provider specialisations are not going to be straight forward for users to implement.

I don't think:

```
someProvider.set { it.dir("child-folder") };
```

Is so onerous that it justifies the glue type specialisations.

**Comment by adammurdoch** [comment]
No-one except us would implement any of these. In general, people would re-use the generic types for their own types. Files and directories get special treatment because they are so common and there's a load of legacy that people have to adapt to and from. Very few other types would get such treatment.

This issue was added based on feedback from trying to use the generic types for dealing with files in the native plugins and native builds. It could probably do with a better title to describe the pain rather than suggest a solution.

**Comment by lacasseio** [comment]
> I don't think:
> 
> `someProvider.set { it.dir("child-folder") };`
> Is so onerous that it justifies the glue type specialisations.

@ldaley Could you expand on this? Are you proposing a different solution? I don't think the code sample you specify is valid with the current API and I'm not sure if we want to pass the current provider/property as an argument to a set closure.

The main issue is the difference between a `Provider` (read-only) and a `Property` (read/write) of a `RegularFile`/`Directory`. Currently, we can only create subdirectory reference from a `Property`. However, in the example, we want to create a subdirectory from a provider. To do so, we have to jump through a lot of hoops.

**Comment by Lanchon** [comment]
maybe this one should be closed in favor of #6390, a close dup.
